plugins/inputs/execd/execd.go

```diff
@@ -0,0 +1,188 @@
+package execd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/kballard/go-shellquote"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+const sampleConfig = `
+  ## Program to run as daemon
+  command = "telegraf-smartctl"
+
+  ## Define how the process is signaled on each collection interval.
+
+  ## Valid values are:
+  ##   "none"   : Do not signal anything.
+  ##              The process must output metrics by itself.
+  ##   "STDIN"  : Send a newline on STDIN.
+  ##   "SIGHUP" : Send a HUP signal. Not available on Windows.
+  signal = "none"
+
+  ## Data format to consume.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+type Execd struct {
+	Command string
+	Signal  string
+
+	acc     telegraf.Accumulator
+	args    []string
+	cmd     *exec.Cmd
+	parser  parsers.Parser
+	stdin   io.WriteCloser
+	stopped bool
+}
+
+func (e *Execd) SampleConfig() string {
+	return sampleConfig
+}
+
+func (e *Execd) Description() string {
+	return "Run executable as long-running input plugin"
+}
+
+func (e *Execd) SetParser(parser parsers.Parser) {
+	e.parser = parser
+}
+
+func (e *Execd) Start(acc telegraf.Accumulator) error {
+	e.acc = acc
+
+	args, err := shellquote.Split(e.Command)
+	if err != nil || len(args) == 0 {
+		return fmt.Errorf("unable to parse command: %s", err)
+	}
+
+	e.args = args
+	go e.cmdRun()
+
+	return nil
+}
+
+func (e *Execd) Stop() {
+	e.stopped = true
+
+	if e.cmd == nil || e.cmd.Process == nil {
```

Comment:
There is a race condition here, the process could start at any time during this function or even after the function exits.  What I suggest doing is to setup a WaitGroup and Context in Start() for a goroutine running the command in a loop, then in stop just call the CancelFunc and wait on the WaitGroup.

Reply:
Fixed in 0e67fe56. The process is now managed in `runLoop()`. A cancelation context is created in `Start()` and used to stop the process. The termination itself is handled in `runLoop()`.

It seems to work but I'd really like to get an additional review here.